### PR TITLE
Fix WeatherDataset boundary checks for analysis indexing and forecast forcing horizon (Fixes #311, Closes #319)

### LINKED ISSUE #311: [BUG] WeatherDataset undercounts __len__ in analysis mode and silently yields NaN tensors out-of-bounds
## Problem

While checking `WeatherDataset` indexing behavior for multi-member ensemble slicing, I found that analysis-mode datasets can be indexed beyond `len(dataset) - 1` without raising `IndexError`.

In those out-of-bounds cases, `__getitem__` may still return tensors, but with malformed content (e.g. truncated forcing windows or NaNs), instead of failing fast.

## Reproduction

```python
from tests.dummy_datastore import DummyDatastore
from neural_lam.weather_dataset import WeatherDataset
import torch

ds = WeatherDataset(
    datastore=DummyDatastore(n_timesteps=10),
    split="train",
    ar_steps=2,
    num_past_forcing_steps=1,
    num_future_forcing_steps=1,
)

print("len:", len(ds))  # currently 5

# Expected: IndexError for both
sample_5 = ds[5]  # returns a sample
sample_6 = ds[6]  # returns sample with malformed forcing

forcing_6 = sample_6[2]
print("forcing shape:", forcing_6.shape)
print("forcing NaN count:", int(torch.isnan(forcing_6).sum())) 
```

## Expected Behavior

- dataset[len(dataset)] and larger indices should raise IndexError.
- Out-of-range access should never produce partially valid samples.

## Actual Behavior

- Some out-of-range indices return samples.
- Forcing tensors can contain NaNs / inconsistent windows.

## Suspected Cause

- Analysis-mode `__len__` boundary appears off-by-one.
- `__getitem__` does not perform explicit index bounds checking.

## Proposed Fix

- Revisit analysis-mode length formula.
- Add explicit bounds checks in __getitem__ (including Python-style negative index handling).
- Add regression tests for out-of-range access.

### LINKED ISSUE #319: [Bug] Forecast-mode WeatherDataset length check ignores forcing window, causing truncated/NaN forcing tensors
## Problem
In forecast mode, `WeatherDataset.__len__` only validates that:
`n_forecast_steps >= 2 + ar_steps`.

However, `_slice_forcing_time` and `_slice_state_time` use:
- `offset = max(2, num_past_forcing_steps)`
- forcing windows that include `num_future_forcing_steps`

So required forecast length is larger when past/future forcing windows are > 1.

## Why this is critical
If forecast horizon is insufficient for the requested forcing window, slicing can be truncated near the end.
This can produce inconsistent forcing window lengths and NaN-padded tensors after concat, silently affecting training/evaluation quality.

## Expected behavior
Dataset initialization should fail fast with a clear `ValueError` when forecast horizon is too short for:
- `ar_steps`
- `num_past_forcing_steps`
- `num_future_forcing_steps`

## Proposed fix
In forecast mode, validate with a stricter requirement (example):
`n_forecast_steps >= max(2, num_past_forcing_steps) + ar_steps + num_future_forcing_steps`

Also add regression tests for forecast mode covering:
- large `num_past_forcing_steps`
- large `num_future_forcing_steps`
- boundary case exactly equal to required horizon
- one case below threshold (must raise ValueError)

### PR DESCRIPTION
**Pull request by kshirajahere**
## Describe your changes

- Fix analysis-mode sample counting in `WeatherDataset.__len__` (off-by-one correction).
- Add explicit index bounds handling in `WeatherDataset.__getitem__`:
  - supports Python-style negative indexing,
  - raises `IndexError` for out-of-range indices.
- Add forecast-mode horizon validation in `WeatherDataset.__len__` so forcing windows cannot overrun available forecast steps.
- Add focused regression tests for:
  - analysis-mode out-of-bounds indexing,
  - forecast-mode forcing horizon boundary checks.
- Add CHANGELOG entry under `## [unreleased] -> ### Fixed`.

## Issue Link

Fixes #311  
Closes #319

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by kshirajahere** [comment]
Ready for review. Happy to make any requested adjustments.

**Comment by kshirajahere** [comment]
Update: I’ve integrated the related forecast-boundary validation from #319 into this PR, so `WeatherDataset` boundary checks are handled in one place.

This PR now includes:
- analysis-mode indexing bounds (`IndexError` behavior),
- forecast-mode forcing-horizon validation in `__len__`,
- regression tests for both paths,
- CHANGELOG update.

@Sadamov @Jayant-kernel please let me know if you’d prefer this split, otherwise this PR should close both #311 and #319 on merge.

**Comment by kshirajahere** [comment]
@sadamov @leifdenby can someone pls review the PR :) thanks

**Comment by kshirajahere** [comment]
@sadamov Hey sir :), was just pinging to request u to review when you have time. Have regression tests done. If you need me to perform any cleanups or any another changes let me know :) thanks

**Comment by kshirajahere** [comment]
@sadamov Please if you get time, do review 😸

File: neural_lam/weather_dataset.py [hunk 1]
```diff
@@ -131,16 +131,47 @@ def __len__(self):
...
+                if n_forcing_forecast_steps < required_forcing_steps:
+                    raise ValueError(
+                        "The number of forcing forecast steps available "
+                        f"({n_forcing_forecast_steps}) is less than the "
+                        f"required {required_forcing_steps} "
+                        f"(max(2, num_past_forcing_steps={self.num_past_forcing_steps})"
+                        f" + ar_steps={self.ar_steps} + "
+                        f"num_future_forcing_steps={self.num_future_forcing_steps}) "
+                        "for constructing forcing windows."
+                    )
```
Comment (sadamov) [inline neural_lam/weather_dataset.py]: Both da_state and da_forcing are always built from the same datastore time coordinates, so their sizes are guaranteed to be equal and a separate shape check is not needed. The single forecast-mode check is still necessary, but only because the required minimum size is larger when forcing is present (+ num_future_forcing_steps), not because the arrays could ever differ in size.

The no-forcing path can remain unchanged in behaviour; the with-forcing path should skip the redundant state check and goes straight to the stricter (and sufficient) forcing constraint.

File: neural_lam/weather_dataset.py [hunk 2]
```diff
@@ -468,6 +500,17 @@ def __getitem__(self, idx):
...
+
+        # Match Python sequence semantics for negative indexing.
+        if idx < 0:
+            idx += dataset_len
+        if idx < 0 or idx >= dataset_len:
+            raise IndexError(
+                f"Index {idx} is out of bounds for dataset of size "
+                f"{dataset_len}"
+            )
+
```
Comment (sadamov) [inline neural_lam/weather_dataset.py]: I was at first conflicted about this change, because it only affects the analysis type data, and only if accessed programmatically e.g. from a test. And I thought that it might rather be separate issue.

But, after consideration, I think this fits well within this PR. And knowing that there is more flexibility and robustness needed when we introduce #138 boundary datastores, this is good. Nothing to change here, just some context.

**Review by sadamov — CHANGES_REQUESTED** [review]
@kshirajahere Thanks alot, this has indeed been a bug. And the fact that datasets were silently truncated is certainly not great. I agree with your suggested fixes and only request a reduction on lines of code, because we know that forcing and state share temporal dimensions.

Because this fix has some importance and you introduced 3 new tests I am asking for a second review as well. I prefer a third pair of eyes before merging a bugfix of this size.

**Comment by kshirajahere** [comment]
Hey @sadamov Can u rereview. I have reduced the forecast-mode validation.

File: neural_lam/weather_dataset.py (hunk 1 above)
Comment (kshirajahere) [reply neural_lam/weather_dataset.py]: I removed the redundant separate forcing-size check and now use the shared forecast horizon once:
- no-forcing path keeps the original `2 + ar_steps` behavior
- with-forcing path applies the stricter minimum needed for the full forcing window

Re-ran:
- pytest -q tests/test_datasets.py -k "dataset_length or out_of_bounds or forecast_len"
- ruff check neural_lam/weather_dataset.py tests/test_datasets.py

**Review by sadamov — APPROVED** [review]
Looks good now! Just waiting for @joeloskarsson review and then we can merge

**Comment by kshirajahere** [comment]
@joeloskarsson pinging for review :)

**Comment by kshirajahere** [comment]
Was going thru the code after pinging @joeloskarsson for the review on slack 😅 and found some gaps (fixed them)
@sadamov small heads-up: I pushed a follow-up fix after rechecking the branch, covering two forecast-mode validation gaps plus regression tests. Since this is post-approval in the same area, I wanted to flag it here.

**Comment by kshirajahere** [comment]
@joeloskarsson i request you to please review my PR :)

**Review by sadamov — APPROVED** [review]
your latest commits did not break anything since my last review. thanks for notifying me

**Comment by kshirajahere** [comment]
Thanks @sadamov, just waiting for @joeloskarsson review so i can make those changes and we can merge this 🚀

**Comment by kshirajahere** [comment]
Heads-up: the two GPU checks are stuck in queue again (Cirun runner availability). Everything else is green. Could a maintainer re-run checks once the Cirun pool is free? If you prefer, I can re-run them too once the runner frees up.

**Comment by kshirajahere** [comment]
@joeloskarsson Would be grateful if you can have a review on this 🙏🏻

**Comment by sadamov** [comment]
Thanks @kshirajahere for the patience on this one. I've rebased onto current main (the original diff predated #239 so the standardization paths it touched were gone) and updated `test_dataset_length` to the corrected off-by-one expectations. Merging now.
